prereise/gather/griddata/hifld/data_process/topology.py

```diff
@@ -217,3 +218,100 @@ def get_mst_edges(lines, substations, **kwargs):
+        components don't match the length of the ``interconnect_size_rank``.
+    """
+    # Create a graph of the network, and drop lines based on substations
+    dropped_lines = lines.loc[
+        lines["SUB_1_ID"].isin(seams_substations)
+        | lines["SUB_2_ID"].isin(seams_substations)
+    ]
+    g = nx.convert_matrix.from_pandas_edgelist(lines, "SUB_1_ID", "SUB_2_ID")
+    g.remove_nodes_from(seams_substations)
+    # Label interconnections based on their sizes
+    sorted_interconnects = sorted(nx.connected_components(g), key=len)[::-1]
+    labels = pd.Series("unknown", index=lines.index, dtype="string")
+    for i, name in enumerate(interconnect_size_rank):
+        labels.loc[lines.SUB_1_ID.isin(sorted_interconnects[i])] = name
+    labels.loc[dropped_lines.index] = "dropped"
+    # Label some dropped lines and unknown interconnection lines (small islands)
+    for interconnect, sub_ids in substation_assumptions.items():
+        labels.loc[
+            lines["SUB_1_ID"].isin(sub_ids) | lines["SUB_2_ID"].isin(sub_ids),
+        ] = interconnect
+    for interconnect, line_ids in line_assumptions.items():
+        labels.loc[line_ids] = interconnect
+    # Use neighboring lines at non-dropped substations to infer line interconnections
+    dropped_lines = lines.loc[labels == "dropped"]
+    non_dropped_lines = lines.loc[labels != "dropped"]
+    for id, line in dropped_lines.iterrows():
+        non_dropped_sub = (  # noqa: F841
+            line.SUB_1_ID if line.SUB_2_ID in seams_substations else line.SUB_2_ID
+        )
+        other_lines = non_dropped_lines.query(
+            "SUB_1_ID == @non_dropped_sub or SUB_2_ID == @non_dropped_sub"
+        )
+        other_line_interconnects = labels.loc[other_lines.index].unique()
+        if len(other_line_interconnects) != 1:
+            raise ValueError(f"Couldn't infer interconnection for line {id}")
+        labels.loc[id] = other_line_interconnects[0]
+    lines["interconnect"] = labels
+
+    # When lines of multiple interconnections meet at a substation, split it
+    for sub_id in sorted(seams_substations):
+        # Find all lines disconnected by removing this substation
+        sub_dropped_lines = lines.query("SUB_1_ID == @sub_id or SUB_2_ID == @sub_id")
+        # Find interconnects for dropped lines which have been successfully labeled
+        new_sub_interconnects = sub_dropped_lines["interconnect"].unique()
+        # Build new substations to replace the old one
+        first_new_sub_id = substations.index.max() + 1
+        new_substations = pd.concat(
+            [substations.loc[sub_id]] * len(new_sub_interconnects), axis=1
+        ).T
+        new_substations.index = pd.RangeIndex(
+            first_new_sub_id, first_new_sub_id + len(new_sub_interconnects)
+        )
+        new_substations["NAME"] = [
+            substations.loc[sub_id, "NAME"] + f"_{i}" for i in new_sub_interconnects
+        ]
```

Comment:
Keeping track of this using `NAME` field is nice, not sure whether it is more straight forward to inferring this using `id` directly, i.e. `xxxx_1` and `xxxx_2` where `xxxx` is the old sub id and the numbers are interconnect rankings. It is not necessarily to have consecutive numbers sub_ids right? We might need them to be integers though.

Reply:
`We might need them to be integers though`: I think you might be right, I'm not sure how much downstream code is implicitly expecting integers vs. just some sort of unique index.